Allow PKCS8 EC private keys to be loaded

tsaarni opened the pull request:
This change makes it possible to use EC (elliptic curve) private keys as client credentials.

Depends on https://github.com/jruby/jruby-openssl/pull/267. Without that fix being applied, this change has no effect.

tsaarni commented:
Tests do not work anymore with current logstash version because of #51.

andsel commented:
@tsaarni please rebase this to `main` to align with latest changes 🙏

tsaarni commented:
> @tsaarni please rebase this to `main` to align with latest changes 🙏

Done!

andsel commented on lib/logstash/outputs/syslog.rb:
```diff
@@ -238,7 +238,7 @@ def setup_ssl
     ssl_context = OpenSSL::SSL::SSLContext.new
     ssl_context.cert = OpenSSL::X509::Certificate.new(File.read(@ssl_cert))
-    ssl_context.key = OpenSSL::PKey::RSA.new(File.read(@ssl_key),@ssl_key_passphrase)
+    ssl_context.key = OpenSSL::PKey::read(File.read(@ssl_key),@ssl_key_passphrase)
```
@Given that you have added all the TLS test scaffolding  in #62  do you think that, after the merge on `main` branch of that PR and rebase of this, could we add 2 tests here, one for RSA key and the other for Elliptic Cryptography here?

tsaarni replied:
Yes absolutely, let's do that.

tsaarni replied:
I needed to remove test `invalid client private key` that was previously added in TLS tests: I now realized that `OpenSSL::PKey::read()` does not raise error when given invalid PEM file, unlike `OpenSSL::PKey::RSA.new` did. Instead, it returns instance of `OpenSSL::PKey::RSA` even though `invalid.pem` clearly is not an RSA key. 

When I fixed support for EC pkey to `PKey::read()` in jruby-openssl, I did not check this. Maybe it lacks negative checks.

andsel replied:
I think that it worthwhile to open an issue in https://github.com/jruby/jruby-openssl repository, accepting silently an improper file is misleading.

andsel replied:
Maybe the difference is that the `read` method instantiate the RSA key 
https://github.com/jruby/jruby-openssl/blob/19135259f121f9c9e95ee7f0b4c830f9267680e0/src/main/java/org/jruby/ext/openssl/PKey.java#L124

instead of the `initiliaze` method 
https://github.com/jruby/jruby-openssl/blob/19135259f121f9c9e95ee7f0b4c830f9267680e0/src/main/java/org/jruby/ext/openssl/PKeyRSA.java#L223

tsaarni replied:
In my test case I used input that does not even have valid PEM block inside, so I would have expected that even PEM block parsing should fail here https://github.com/jruby/jruby-openssl/blob/19135259f121f9c9e95ee7f0b4c830f9267680e0/src/main/java/org/jruby/ext/openssl/PKey.java#L117  -> 
https://github.com/jruby/jruby-openssl/blob/19135259f121f9c9e95ee7f0b4c830f9267680e0/src/main/java/org/jruby/ext/openssl/x509store/PEMInputOutput.java#L317
Requires some additional debugging on jruby-openssl to find out what is going on there...

I've created new issue https://github.com/jruby/jruby-openssl/issues/285. I added also the CA case which I noticed earlier.

andsel replied:
Oks, let's see what they answer and then I'll merge this.

tsaarni replied:
Jruby-openssl function `PKey::read()` was fixed in release https://github.com/jruby/jruby-openssl/releases/tag/v0.14.4  Apr 11, 2024.  See https://github.com/jruby/jruby-openssl/commit/ab355cad013d1da2637439aeadda43924a85c7c7. 

It now returns error when reading invalid key file so I've restored the test case "invalid client certificate".

andsel reviewed (commented):
Left a comment given that #62 add TLS test harness.

tsaarni commented:
Hi @andsel The issue with jruby-openssl is now sorted out https://github.com/logstash-plugins/logstash-output-syslog/pull/61#discussion_r1980889928. I've updated and rebased the PR, it should be ready for merging.

tsaarni commented:
I’ve rebased this PR and squashed it into a single commit to remove now-obsolete cruft gathered during the lifetime of this PR.

Cc @mashhurs
